File: packages/shared/src/subscription-renewal.ts
import { Temporal } from "@js-temporal/polyfill";
import {
  type BillingCycle,
  type CustomCycleUnit,
  type DateOnly,
  type SubscriptionStatus,
  isValidDateOnly,
} from "./runtime";

/** 续订模式决定推进阈值：自动维护追到 today，手动续订至少推进一期并严格晚于当前边界。 */
export type RenewalMode = "auto" | "manual";

/** 续订算法输入是跨 D1、PocketBase 和前端 fixture 的最小字段集，不包含价格、通知或展示字段。 */
export interface SubscriptionRenewalInput {
  billingCycle: BillingCycle;
  status: SubscriptionStatus;
  startDate: string;
  nextBillingDate: string;
  autoRenew: boolean;
  autoCalculateNextBillingDate: boolean;
  customDays?: number | null | undefined;
  customCycleUnit?: CustomCycleUnit | null | undefined;
}

/** 账单日纯计算入口使用同一字段集，避免表单自动日期与后端续订算法分叉。 */
export interface AdvanceBillingDateInput {
  billingCycle: BillingCycle;
  startDate: string;
  nextBillingDate: string;
  autoCalculateNextBillingDate: boolean;
  customDays?: number | null | undefined;
  customCycleUnit?: CustomCycleUnit | null | undefined;
}

/** 续订结果只返回 date-only 与状态；不会生成付款记录或通知历史。 */
export interface SubscriptionRenewalResult {
  nextBillingDate: DateOnly;
  status: SubscriptionStatus;
}

const MAX_ADVANCE_CYCLES = 20_000;

/**
 * 判断订阅是否可由后台维护任务自动推进。
 *
 * 自动续订只处理已经落后于用户本地 today 的 active/trial 周期订阅；缺省 autoRenew 不能被解释成授权。
 */
export function isAutoRenewEligible(subscription: SubscriptionRenewalInput, today: string): boolean {
  return (
    subscription.autoRenew &&
    subscription.billingCycle !== "one-time" &&
    (subscription.status === "active" || subscription.status === "trial") &&
    isValidDateOnly(subscription.startDate) &&
    isValidDateOnly(subscription.nextBillingDate) &&
    isValidDateOnly(today) &&
    compareDateOnly(subscription.nextBillingDate, today) < 0
  );
}

/**
 * 判断订阅是否可由用户手动续订。
 *
 * 手动续订覆盖 expired 记录，但明确排除 autoRenew=true 的订阅，避免用户和维护 cron 同时推进同一账单日。
 */
export function isManualRenewEligible(subscription: SubscriptionRenewalInput): boolean {
  return (
    !subscription.autoRenew &&
    subscription.billingCycle !== "one-time" &&
    (subscription.status === "active" || subscription.status === "trial" || subscription.status === "expired") &&
    isValidDateOnly(subscription.startDate) &&
    isValidDateOnly(subscription.nextBillingDate)
  );
}

/**
 * 推进订阅续订状态，是 Docker Go、Cloudflare Worker 和前端测试共用的事实算法。
 *
 * `mode=auto` 推进到第一个 `>= today` 的周期日；`mode=manual` 至少推进一期并要求结果严格晚于阈值。
 */
export function advanceSubscriptionRenewal(
  subscription: SubscriptionRenewalInput,
  today: string,
  mode: RenewalMode,
): SubscriptionRenewalResult | null {
  if (mode === "auto" && !isAutoRenewEligible(subscription, today)) return null;
  if (mode === "manual" && !isManualRenewEligible(subscription)) return null;
  const nextBillingDate = advanceBillingDate(subscription, today, mode);
  return {
    nextBillingDate,
    status: mode === "manual" && subscription.status === "expired" ? "active" : subscription.status,
  };
}

/**
 * 计算下一账单日，不改变状态。
 *
 * `autoCalculateNextBillingDate=true` 以 startDate 作周期锚点；否则保留用户手动修正过的 nextBillingDate 锚点。
 */
export function advanceBillingDate(
  input: AdvanceBillingDateInput,
  today: string,
  mode: RenewalMode,
): DateOnly {
  assertRenewableBillingCycle(input.billingCycle);
  const original = assertDateOnly(input.nextBillingDate);
  const anchor = assertDateOnly(input.autoCalculateNextBillingDate ? input.startDate : input.nextBillingDate);
  const threshold = mode === "manual" && compareDateOnly(original, today) > 0 ? original : assertDateOnly(today);
  const strict = mode === "manual";

  return firstCycleDateAfter(anchor, input, threshold, strict);
}

/** 表单自动推算下一账单日的纯函数入口，保持 date-only 输出，不引入浏览器时区。 */
export function calculateNextBillingDate(
  startDate: string,
  cycle: BillingCycle,
  customDays?: number | null | undefined,
  referenceDate?: string | null | undefined,
  customCycleUnit: CustomCycleUnit = "day",
): DateOnly {
  const anchor = assertDateOnly(startDate);
  if (cycle === "one-time") return anchor;
  const threshold = referenceDate ? assertDateOnly(referenceDate) : anchor;
  return firstCycleDateAfter(anchor, {
    billingCycle: cycle,
    startDate: anchor,
    nextBillingDate: anchor,
    autoCalculateNextBillingDate: true,
    customDays,
    customCycleUnit,
  }, threshold, false);
}

/**
 * 将一个 date-only 按账单周期前进 N 期。
 *
 * 使用 Temporal 是为了让月末夹取语义稳定，例如 1 月 31 日按月推进到 2 月最后一天。
 */
export function addBillingCycles(
  date: string,
  cycle: BillingCycle,
  cycleCount: number,
  customDays?: number | null | undefined,
  customCycleUnit: CustomCycleUnit = "day",
): DateOnly {
  const start = toPlainDate(date);
  const count = Math.max(1, Math.trunc(cycleCount));
  const customCount = Math.max(1, Math.trunc(customDays ?? 30)) * count;
  switch (cycle) {
    case "weekly":
      return fromPlainDate(start.add({ weeks: count }));
    case "monthly":
      return fromPlainDate(start.add({ months: count }));
    case "quarterly":
      return fromPlainDate(start.add({ months: 3 * count }));
    case "semi-annual":
      return fromPlainDate(start.add({ months: 6 * count }));
    case "annual":
      return fromPlainDate(start.add({ years: count }));
    case "custom":
      return addCustomBillingCycles(start, customCount, customCycleUnit);
    case "one-time":
      return fromPlainDate(start);
  }
}

function firstCycleDateAfter(
  anchor: string,
  input: AdvanceBillingDateInput,
  threshold: string,
  strict: boolean,
): DateOnly {
  const initialCycles = initialCycleCount(anchor, input, threshold, strict);
  let cycleCount = Math.max(1, initialCycles);
  for (let attempts = 0; attempts < MAX_ADVANCE_CYCLES; attempts += 1) {
    const candidate = addBillingCycles(anchor, input.billingCycle, cycleCount, input.customDays, input.customCycleUnit ?? "day");
    const comparison = compareDateOnly(candidate, threshold);
    if (strict ? comparison > 0 : comparison >= 0) return candidate;
    cycleCount += 1;
  }
  // 保护异常自定义周期或脏数据，避免维护任务在单条订阅上无限循环占满 Worker/Go cron。
  throw new Error("SUBSCRIPTION_RENEWAL_ADVANCE_LIMIT_EXCEEDED");
}

function initialCycleCount(
  anchor: string,
  input: AdvanceBillingDateInput,
  threshold: string,
  strict: boolean,
): number {
  const dayStep = exactDayStep(input);
  if (!dayStep) return 1;
  // 只有“固定天数”周期能直接跳到接近阈值的期数；月份/年份必须逐期推进以保留月末夹取语义。
  const diff = toPlainDate(anchor).until(toPlainDate(threshold), { largestUnit: "day" }).days;
  const adjusted = strict ? diff + 1 : diff;
  return Math.max(1, Math.ceil(adjusted / dayStep));
}

function exactDayStep(input: Pick<AdvanceBillingDateInput, "billingCycle" | "customDays" | "customCycleUnit">): number | null {
  if (input.billingCycle === "weekly") return 7;
  if (input.billingCycle !== "custom") return null;
  const count = Math.max(1, Math.trunc(input.customDays ?? 30));
  if ((input.customCycleUnit ?? "day") === "day") return count;
  if (input.customCycleUnit === "week") return count * 7;
  return null;
}

function addCustomBillingCycles(
  start: Temporal.PlainDate,
  count: number,
  unit: CustomCycleUnit,
): DateOnly {
  switch (unit) {
    case "week":
      return fromPlainDate(start.add({ weeks: count }));
    case "month":
      return fromPlainDate(start.add({ months: count }));
    case "year":
      return fromPlainDate(start.add({ years: count }));
    case "day":
      return fromPlainDate(start.add({ days: count }));
  }
}

function assertRenewableBillingCycle(cycle: BillingCycle): asserts cycle is Exclude<BillingCycle, "one-time"> {
  if (cycle === "one-time") {
    throw new Error("SUBSCRIPTION_RENEWAL_ONE_TIME_NOT_RENEWABLE");
  }
}

function assertDateOnly(value: string): DateOnly {
  if (!isValidDateOnly(value)) {
    throw new Error(`Invalid date-only value: ${value}`);
  }
  return value as DateOnly;
}

function toPlainDate(value: string): Temporal.PlainDate {
  return Temporal.PlainDate.from(assertDateOnly(value));
}

function fromPlainDate(value: Temporal.PlainDate): DateOnly {
  return assertDateOnly(value.toString());
}

function compareDateOnly(left: string, right: string): number {
  return Temporal.PlainDate.compare(toPlainDate(left), toPlainDate(right));
}

File: packages/server/cmd/renewlet/subscription_renewal.go
package main

// subscription_renewal.go 是 Go/PocketBase 运行面的订阅续订算法。
//
// 架构位置：shared TypeScript 版本是事实源，Go 版通过同一组 JSON fixture 对齐，
// 供 Docker cron、通知生成前维护和手动续订 API 使用。
//
// 注意：所有输入输出都是 date-only 字符串，不引入服务器本地时区；today 由调用方按用户时区生成。
import (
	"errors"
	"math"
	"strings"
	"time"
)

type renewalMode string

const (
	renewalModeAuto   renewalMode = "auto"
	renewalModeManual renewalMode = "manual"
	maxAdvanceCycles              = 20000
)

type subscriptionRenewalInput struct {
	BillingCycle                 string
	Status                       string
	StartDate                    string
	NextBillingDate              string
	AutoRenew                    bool
	AutoCalculateNextBillingDate bool
	CustomDays                   int
	CustomCycleUnit              string
}

type subscriptionRenewalResult struct {
	NextBillingDate string
	Status          string
}

func subscriptionRenewalInputFromRecord(record subscriptionRecordReader) subscriptionRenewalInput {
	return subscriptionRenewalInput{
		BillingCycle:                 record.GetString("billingCycle"),
		Status:                       record.GetString("status"),
		StartDate:                    record.GetString("startDate"),
		NextBillingDate:              record.GetString("nextBillingDate"),
		AutoRenew:                    record.GetBool("autoRenew"),
		AutoCalculateNextBillingDate: record.GetBool("autoCalculateNextBillingDate"),
		CustomDays:                   record.GetInt("customDays"),
		CustomCycleUnit:              record.GetString("customCycleUnit"),
	}
}

type subscriptionRecordReader interface {
	GetString(string) string
	GetBool(string) bool
	GetInt(string) int
}

func isAutoRenewEligible(input subscriptionRenewalInput, today string) bool {
	// 自动续订只处理已经落后于用户本地 today 的 active/trial 周期订阅，缺省 false 不能被解释成授权。
	return input.AutoRenew &&
		input.BillingCycle != "one-time" &&
		(input.Status == "active" || input.Status == "trial") &&
		isValidDateOnly(input.StartDate) &&
		isValidDateOnly(input.NextBillingDate) &&
		isValidDateOnly(today) &&
		input.NextBillingDate < today
}

func isManualRenewEligible(input subscriptionRenewalInput) bool {
	// 手动续订允许 expired 重新回到 active，但排除 autoRenew=true，避免和维护 cron 同时推进。
	return !input.AutoRenew &&
		input.BillingCycle != "one-time" &&
		(input.Status == "active" || input.Status == "trial" || input.Status == "expired") &&
		isValidDateOnly(input.StartDate) &&
		isValidDateOnly(input.NextBillingDate)
}

func advanceSubscriptionRenewal(input subscriptionRenewalInput, today string, mode renewalMode) (subscriptionRenewalResult, bool, error) {
	switch mode {
	case renewalModeAuto:
		if !isAutoRenewEligible(input, today) {
			return subscriptionRenewalResult{}, false, nil
		}
	case renewalModeManual:
		if !isManualRenewEligible(input) {
			return subscriptionRenewalResult{}, false, nil
		}
	default:
		return subscriptionRenewalResult{}, false, errors.New("SUBSCRIPTION_RENEWAL_MODE_INVALID")
	}

	nextBillingDate, err := advanceBillingDate(input, today, mode)
	if err != nil {
		return subscriptionRenewalResult{}, false, err
	}
	status := input.Status
	if mode == renewalModeManual && status == "expired" {
		status = "active"
	}
	return subscriptionRenewalResult{NextBillingDate: nextBillingDate, Status: status}, true, nil
}

func advanceBillingDate(input subscriptionRenewalInput, today string, mode renewalMode) (string, error) {
	if input.BillingCycle == "one-time" {
		return "", errors.New("SUBSCRIPTION_RENEWAL_ONE_TIME_NOT_RENEWABLE")
	}
	original, err := parseDateOnly(input.NextBillingDate)
	if err != nil {
		return "", err
	}
	todayDate, err := parseDateOnly(today)
	if err != nil {
		return "", err
	}
	anchorValue := input.NextBillingDate
	if input.AutoCalculateNextBillingDate {
		anchorValue = input.StartDate
	}
	anchor, err := parseDateOnly(anchorValue)
	if err != nil {
		return "", err
	}
	threshold := todayDate
	if mode == renewalModeManual && original.After(todayDate) {
		threshold = original
	}
	// 手动续订 strict=true：即使当前日期还没到，也至少推进一期，防止按钮点击后看起来“没有变化”。
	return firstCycleDateAfter(anchor, input, threshold, mode == renewalModeManual)
}

func firstCycleDateAfter(anchor time.Time, input subscriptionRenewalInput, threshold time.Time, strict bool) (string, error) {
	cycleCount := maxInt(1, initialCycleCount(anchor, input, threshold, strict))
	for attempts := 0; attempts < maxAdvanceCycles; attempts++ {
		candidate, err := addBillingCyclesDate(anchor, input.BillingCycle, cycleCount, input.CustomDays, input.CustomCycleUnit)
		if err != nil {
			return "", err
		}
		if (!strict && !candidate.Before(threshold)) || (strict && candidate.After(threshold)) {
			return formatDateOnly(candidate), nil
		}
		cycleCount++
	}
	// 脏数据或极端自定义周期不能让单条订阅把 cron 卡成无限循环。
	return "", errors.New("SUBSCRIPTION_RENEWAL_ADVANCE_LIMIT_EXCEEDED")
}

func initialCycleCount(anchor time.Time, input subscriptionRenewalInput, threshold time.Time, strict bool) int {
	dayStep := exactDayStep(input)
	if dayStep <= 0 {
		return 1
	}
	// 只有固定天数周期能直接跳到近似期数；月/年周期必须逐期推进以保留月末夹取语义。
	diff := int(threshold.Sub(anchor).Hours() / 24)
	if strict {
		diff++
	}
	return maxInt(1, int(math.Ceil(float64(diff)/float64(dayStep))))
}

func exactDayStep(input subscriptionRenewalInput) int {
	if input.BillingCycle == "weekly" {
		return 7
	}
	if input.BillingCycle != "custom" {
		return 0
	}
	count := maxInt(1, input.CustomDays)
	switch normalizeCustomCycleUnit(input.CustomCycleUnit) {
	case "day":
		return count
	case "week":
		return count * 7
	default:
		return 0
	}
}

func normalizeCustomCycleUnit(value string) string {
	switch strings.TrimSpace(value) {
	case "week", "month", "year":
		return strings.TrimSpace(value)
	default:
		return "day"
	}
}

func addBillingCyclesDate(anchor time.Time, cycle string, cycleCount int, customDays int, customCycleUnit string) (time.Time, error) {
	count := maxInt(1, cycleCount)
	customCount := maxInt(1, customDays)
	switch cycle {
	case "weekly":
		return anchor.AddDate(0, 0, 7*count), nil
	case "monthly":
		return addDateClamped(anchor, 0, count), nil
	case "quarterly":
		return addDateClamped(anchor, 0, 3*count), nil
	case "semi-annual":
		return addDateClamped(anchor, 0, 6*count), nil
	case "annual":
		return addDateClamped(anchor, count, 0), nil
	case "custom":
		return addCustomBillingCyclesDate(anchor, customCount*count, customCycleUnit), nil
	case "one-time":
		return anchor, nil
	default:
		return time.Time{}, errors.New("SUBSCRIPTION_RENEWAL_BILLING_CYCLE_INVALID")
	}
}

func addCustomBillingCyclesDate(anchor time.Time, count int, unit string) time.Time {
	switch normalizeCustomCycleUnit(unit) {
	case "week":
		return anchor.AddDate(0, 0, 7*count)
	case "month":
		return addDateClamped(anchor, 0, count)
	case "year":
		return addDateClamped(anchor, count, 0)
	default:
		return anchor.AddDate(0, 0, count)
	}
}

func addDateClamped(anchor time.Time, years int, months int) time.Time {
	// Go AddDate 会把 1 月 31 日 +1 月规范化到 3 月；续订契约需要和 Temporal 一样夹到目标月最后一天。
	year, month, day := anchor.Date()
	totalMonths := int(month) - 1 + months + years*12
	targetYear := year + totalMonths/12
	targetMonthIndex := totalMonths % 12
	if targetMonthIndex < 0 {
		targetMonthIndex += 12
		targetYear--
	}
	targetMonth := time.Month(targetMonthIndex + 1)
	targetDay := minInt(day, daysInMonth(targetYear, targetMonth))
	return time.Date(targetYear, targetMonth, targetDay, 0, 0, 0, 0, time.UTC)
}

func daysInMonth(year int, month time.Month) int {
	return time.Date(year, month+1, 0, 0, 0, 0, 0, time.UTC).Day()
}

func parseDateOnly(value string) (time.Time, error) {
	value = strings.TrimSpace(value)
	parsed, err := time.Parse("2006-01-02", value)
	if err != nil {
		return time.Time{}, err
	}
	return parsed, nil
}

func formatDateOnly(value time.Time) string {
	return value.UTC().Format("2006-01-02")
}
